in-progress/7346-batch-proving-circuits-and-l1.md

```diff
@@ -0,0 +1,259 @@
+
+- `left.chain_id == right.chain_id`
+- `left.version == right.version`
+- `left.block_number == right.block_number + 1`
+- `left.timestamp < right.timestamp`
+- `coinbase`, `fee_recipient`, and `gas_fees` are not constrained (though `gas_fees` may be in a 1559-like world)
+
+### Root rollup
+
+The new root rollup circuit then takes two `BlockRootOrBlockMergePublicInputs`, performs the same checks as the block merge rollup, but outputs a subset of the public inputs, to make L1 verification cheaper:
+
+```rust
+struct RootRollupPublicInputs {
+  previous_archive: Field,
+  end_archive: Field,
+  end_block_hash: Field,
+  end_timestamp: Field,
+  end_block_number: Field,
+  out_hash: Field,
+  fees: [{ recipient: Address, value: Field }; 32]
+}
+```
+
+### Empty block root rollup
+
+Since we no longer submit a proof per block, and thanks to @MirandaWood we now have wonky rollups, we no longer need to fill a block with "empty" txs. This means we can discard the empty private kernel circuit.
+
+However, we still need to be able to represent an empty block. We can do this by introducing an empty block root rollup circuit, which outputs a `BlockRootOrBlockMergePublicInputs` which does not consume any merge rollups, and the end state just equals the start state. Note that we may still need an empty nested circuit to fill in the nested proof.
+
+## L1 Rollup Contract
+
+The Rollup contract today has a main entrypoint `process(header, archive, aggregationObject, proof)`. We propose breaking this method into two:
+
+```solidity
+contract Rollup {
+  process(header, archive);
+  submitProof(publicInputs, aggregationObject, proof);
+}
+```
+
+### State
+
+To track both the proven and unproven chains, we add the following state variables to the contract:
+
+```diff
+contract Rollup {
+  bytes32 lastArchiveTreeRoot;
+  uint256 lastBlockTimestamp;
++ bytes32 verifiedArchiveTreeRoot;
++ uint256 verifiedBlockTimestamp;
+}
+```
+
+The `last*` fields are updated every time a new block is uploaded, while the `verified*` ones are updated when a proof is uploaded. In the event of a rollback due to failure on proof submission, the `last*` fields are overwritten with the contents of `verified*`.
+
+### Process
+
+Today the `process` method does the following:
+
+1. Validate the new block header
+2. Check data availability against the availability oracle
+3. Update the `lastArchiveTreeRoot` and `lastBlockTimestamp` in the contract storage
+4. Consume L1-to-L2 messages from the Inbox
+5. Emit an `L2BlockProcessed` event with the block number
+6. Verify the root rollup proof
+7. Insert L2-to-L1 messages into the Outbox using the `out_hash` and block number
+8. Pay out `total_fees` to `coinbase` in the L1 gas token
+
+The first five items can keep being carried out by the `process` method for each new block that gets submitted. Proof verification, L2-to-L1 messages, and fee payment messages are moved to `submitProof`.
```

Comment:
> The process name is unclear to me. Would this be better described as submitBlockProposal?

Depends on what we define as a "block proposal". Naming strikes again! If you mean "a block that is part of the unproven chain", then yeah.

> I think perhaps 2 (which will be calling the blob point eval precompile on some earlier-submitted blob) should be deferred until the proof is being verified.

Blobification is not yet clear to me, but yeah, we can defer that until the proof is submitted and verified if it makes sense. FWIW there's a step missing here that is checking the attestations from the validator committee, but that'll come in later.

> The L1-to-L2 messages shouldn't be removed from the L1 box until the proof has been verified

Good point! Do you know if we'd need to soft-delete them at least, in order to prevent them from being consumed in two unproven blocks in a row? Or is there another mechanism to prevent this?

> Validate the new block header - I thought we didn't want to submit potentially malicious state roots at the time of proposing a block? (and hence shouldn't be submitting new block headers to L1 at this stage)

This proposal is assuming we're still uploading tx effects, but yeah, this may change if/when we switch to tx objects.

Reply:
> The L1-to-L2 messages shouldn't be removed from the L1 box until the proof has been verified

Was wondering about the same thing! I think we need to restore `toConsume` (but not `isProgress`) in the inbox to the point `verifiedArchiveTreeRoot` was set when a proof fails.

Was also considering deferring `l1_to_l2_roots` to be proved in the final root rollup and consume the root in the inbox from `submiteProof`. But this will result in a bigger delay for users to make use of a l1_to_l2_msg. So maybe not a good idea.

> This proposal is assuming we're still uploading tx effects, but yeah, this may change if/when we switch to tx objects.

What's tx objects?